govcd/vm_affinity_rule.go

```diff
@@ -0,0 +1,397 @@
+	for _, rule := range fullList {
+		if rule.Polarity == polarity {
+			returnList = append(returnList, rule)
+		}
+	}
+
+	return returnList, nil
+}
+
+// GetVmAffinityRuleByHref finds a VM affinity or anti-affinity rule by HREF
+func (vdc *Vdc) GetVmAffinityRuleByHref(href string) (*VmAffinityRule, error) {
+
+	affinityRule := NewVmAffinityRule(vdc.client)
+
+	_, err := vdc.client.ExecuteRequest(href, http.MethodGet,
+		"", "error retrieving affinity rule: %s", nil, affinityRule.VmAffinityRule)
+	if err != nil {
+		return nil, err
+	}
+
+	//fmt.Printf("<<<%# v>>>\n",pretty.Formatter(affinityRule.VmAffinityRule))
+	return affinityRule, nil
+}
+
+// GetVmAffinityRulesByName finds the rules with the given name
+// Note that name does not have to be unique, so a search by name can match several items
+// If polarity is indicated, the function retrieves only the rules with the given polarity
+func (vdc *Vdc) GetVmAffinityRulesByName(name string, polarity string) ([]*VmAffinityRule, error) {
+
+	var returnList []*VmAffinityRule
+	ruleList, err := vdc.GetAllVmAffinityRuleList()
+	if err != nil {
+		return nil, err
+	}
+	for _, rule := range ruleList {
+		if rule.Name == name {
+			fullRule, err := vdc.GetVmAffinityRuleByHref(rule.HREF)
+			if err != nil {
+				return returnList, err
+			}
+			if (polarity != "" && polarity == rule.Polarity) || polarity == "" {
+				returnList = append(returnList, fullRule)
+			}
+		}
+	}
+	return returnList, nil
+}
+
+// GetVmAffinityRuleById retrieves a VM affinity or anti-affinity rule by ID
+func (vdc *Vdc) GetVmAffinityRuleById(id string) (*VmAffinityRule, error) {
+
+	list, err := vdc.GetAllVmAffinityRuleList()
+	if err != nil {
+		return nil, err
+	}
+	for _, rule := range list {
+		if equalIds(id, rule.ID, rule.HREF) {
+			return vdc.GetVmAffinityRuleByHref(rule.HREF)
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetVmAffinityRuleByNameOrId retrieves an affinity or anti-affinity rule by name or ID
+// Given the possibility of a name identifying multiple items, this function may also fail
+// when the search by name returns more than one item.
+func (vdc *Vdc) GetVmAffinityRuleByNameOrId(identifier string) (*VmAffinityRule, error) {
+	getByName := func(name string, refresh bool) (interface{}, error) {
+		list, err := vdc.GetVmAffinityRulesByName(name, "")
+		if err != nil {
+			return nil, err
+		}
+		if len(list) == 0 {
+			return nil, ErrorEntityNotFound
+		}
+		if len(list) == 1 {
+			return list[0], nil
+		}
+		return nil, fmt.Errorf("more than item matches the name '%s'", name)
+	}
+	getById := func(id string, refresh bool) (interface{}, error) { return vdc.GetVmAffinityRuleById(id) }
+	entity, err := getEntityByNameOrId(getByName, getById, identifier, false)
+	if entity == nil {
+		return nil, err
+	}
+	return entity.(*VmAffinityRule), err
+}
+
+func validateAffinityRule(client *Client, affinityRuleDef *types.VmAffinityRule, checkVMs bool) (*types.VmAffinityRule, error) {
```

Comment:
comment on function would be nice which explains checkVMs :)

Reply:
done